ipam/aws: fixed a bug causing the operator to hang indefinitely when the ENI limits for an instance type could not be determined

mvisonneau opened the pull request:
Fixes: #14802

With the current implementation, there is a race condition which causes the operator to completely hang if one of the EC2 instance of the cluster has a type for which the operator cannot determine the limits. This PR sorts this situation by removing the attempt for getting a readlock over the node mutex which already has a readwrite lock at this time.

Here is the pprof trace of the hanging goroutine:
```
1 @ 0x43a4a5 0x44afc5 0x44afae 0x46c0e7 0x25c769a 0x25c75e2 0x2649a7b 0x264da08 0x25c7302 0x25c7a6c 0x25c7758 0x25ce2bc 0x25cc525 0x25cc375 0x2660af1 0x19ec99a 0x1b1b61f 0x19dece2 0x19dc3a2 0x1116ebf 0x1115ced 0x1115c18 0x19dc105 0x19dc0a6 0x46fe81
#	0x46c0e6	sync.runtime_SemacquireMutex+0x46						/usr/local/go/src/runtime/sema.go:71
#	0x25c7699	sync.(*RWMutex).RLock+0xd9							/usr/local/go/src/sync/rwmutex.go:50
#	0x25c75e1	github.com/cilium/cilium/pkg/ipam.(*Node).InstanceID+0x21			/go/src/github.com/cilium/cilium/pkg/ipam/node.go:311
#	0x2649a7a	github.com/cilium/cilium/pkg/aws/eni.(*Node).loggerLocked+0x7a			/go/src/github.com/cilium/cilium/pkg/aws/eni/node.go:76
#	0x264da07	github.com/cilium/cilium/pkg/aws/eni.(*Node).GetMaximumAllocatableIPv4+0x67	/go/src/github.com/cilium/cilium/pkg/aws/eni/node.go:503
#	0x25c7301	github.com/cilium/cilium/pkg/ipam.(*Node).getMaxAllocate+0x41			/go/src/github.com/cilium/cilium/pkg/ipam/node.go:222
#	0x25c7a6b	github.com/cilium/cilium/pkg/ipam.(*Node).recalculate+0x18b			/go/src/github.com/cilium/cilium/pkg/ipam/node.go:378
#	0x25c7757	github.com/cilium/cilium/pkg/ipam.(*Node).UpdatedResource+0x97			/go/src/github.com/cilium/cilium/pkg/ipam/node.go:337
#	0x25ce2bb	github.com/cilium/cilium/pkg/ipam.(*NodeManager).Update.func1+0x5b		/go/src/github.com/cilium/cilium/pkg/ipam/node_manager.go:263
#	0x25cc524	github.com/cilium/cilium/pkg/ipam.(*NodeManager).Update+0x184			/go/src/github.com/cilium/cilium/pkg/ipam/node_manager.go:321
#	0x25cc374	github.com/cilium/cilium/pkg/ipam.(*NodeManager).Create+0x34			/go/src/github.com/cilium/cilium/pkg/ipam/node_manager.go:251
#	0x2660af0	main.startSynchronizingCiliumNodes.func1+0x70					/go/src/github.com/cilium/cilium/operator/cilium_node.go:59
#	0x19ec999	k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnAdd+0x59		/go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:231
#	0x1b1b61e	github.com/cilium/cilium/pkg/k8s/informer.NewInformerWithStore.func1+0x25e	/go/src/github.com/cilium/cilium/pkg/k8s/informer/informer.go:119
```

As we discussed with @christarazi in the releated issue, I also took this opportunity to improve the associated logging!

```release-note
ipam/aws: fixed a bug causing the operator to hang indefinitely when the ENI limits for an instance type could not be determined
```

christarazi commented on pkg/aws/eni/node.go:
```diff
@@ -174,7 +174,7 @@ func (n *Node) ReleaseIPs(ctx context.Context, r *ipam.ReleaseAction) error {
 	limits, limitsAvailable := n.getLimits()
 	if !limitsAvailable {
-		return nil, fmt.Errorf("Unable to determine limits")
+		return nil, fmt.Errorf(errUnableToDetermineLimits)
```
I think moving to sentinel error values is fine, but no need to `fmt.Errorf()` again

mvisonneau replied:
sounds good, what should I use then as `errUnableToDetermineLimits` is a string? 🤔

christarazi replied:
Yeah you can do:

```
var errUnableToDetermineLimits = errors.New("...")
```

Then 
```suggestion
		return nil, errUnableToDetermineLimits
```

mvisonneau replied:
I reverted those changes as they are not related to this change anyway!

christarazi commented on pkg/aws/eni/node.go:
```diff
@@ -322,12 +322,13 @@ func (n *Node) findNextIndex(index int32) int32 {
+	errUnableToAttachENI                    = "unable to attach ENI"
+	errUnableToCreateENI                    = "unable to create ENI"
+	errUnableToDetermineFirstInterfaceIndex = "unable to determine first interface index"
+	errUnableToDetermineLimits              = "unable to determine limits, consider setting --update-ec2-adapter-limit-via-api=true"
```
I think we should keep recommendations / mitigations to log msgs instead of error msgs. Errors should describe succinctly the cause of a problem, but it should not recommend any course of action. I believe it is cleaner if the caller of the function which errored to recommend a course of action / provide a mitigation instead.  This way the caller has the full context of the issue and can decide if it makes sense to log at that point in the code path.

mvisonneau replied:
sounds good, I will revert this change then 👍, where exactly do you reckon we should be putting the advice?

christarazi replied:
I think the first place that the user will encounter this error is in `pkg/nodediscovery.determineFirstInterfaceIndex()` so it would make sense to put it there to warn as early as possible

mvisonneau replied:
sounds great, I added the warning at this place 👍

christarazi commented on pkg/aws/eni/node.go:
```diff
@@ -494,9 +495,9 @@ func (n *Node) ResyncInterfacesAndIPs(ctx context.Context, scopedLog *logrus.Ent
+func (n *Node) GetMaximumAllocatableIPv4(scopedLog *logrus.Entry) int {
 	if n == nil {
-		log.Warningf("Could not determine first interface index, %s", getMaximumAllocatableIPv4FailureWarningStr)
+		log.Warningf("%s: %s", errUnableToDetermineFirstInterfaceIndex, getMaximumAllocatableIPv4FailureWarningStr)
```
This seems unnecessary to me. I think we should keep the log msgs how they were, as log msgs and error msgs should not be used interchangeably.

mvisonneau replied:
sounds good I will revert these then 👍

mvisonneau replied:
done 👍

christarazi commented on pkg/aws/eni/node.go:
```diff
@@ -494,9 +495,9 @@ func (n *Node) ResyncInterfacesAndIPs(ctx context.Context, scopedLog *logrus.Ent
 // GetMaximumAllocatableIPv4 returns the maximum amount of IPv4 addresses
 // that can be allocated to the instance
-func (n *Node) GetMaximumAllocatableIPv4() int {
+func (n *Node) GetMaximumAllocatableIPv4(scopedLog *logrus.Entry) int {
```
Why was this needed? It seems to me that we can already access the logger via `n.loggerLocked()`

christarazi replied:
Plus this modifies the interface to the functions for no concrete benefit

mvisonneau replied:
So this is exactly where the whole situation resides. By calling `n.loggerLocked()` within this scope, the function gets blocked as it returns `n.node.InstanceID()` which requires a read-lock over the node mutex which is already rw locked

https://github.com/cilium/cilium/blob/e2676483a51b7b089cd6a476b83038b9a1ec6169/pkg/aws/eni/node.go#L76

https://github.com/cilium/cilium/blob/e2676483a51b7b089cd6a476b83038b9a1ec6169/pkg/ipam/node.go#L311

https://github.com/cilium/cilium/blob/e2676483a51b7b089cd6a476b83038b9a1ec6169/pkg/ipam/node.go#L364

do you know if there is a better way to workaround this issue?

christarazi replied:
Are you saying that you found a deadlock in the code prior and this change fixes it? Because the prior code is still using `loggerLocked()`

mvisonneau replied:
yes, absolutely!

christarazi replied:
Ah, I see! I didn't notice that the driver behind this PR is this deadlock. Nice catch!

However, I still think we can do better here than to pollute the interface with a logger; that shouldn't need to be passed around and the fact that we do rely on that to avoid a deadlock is a very strong indicator that we need to rethink the flow and dependencies of this code.

Here's my proposal:

Let's add `instanceID string` as a field to the `pkg/aws/eni.Node` struct. This would allow us to avoid depending on `n.node.InstanceID()` which was done previously (which evidently cannot be called when `n.node` is already locked).

We can initialize the field when we call [`CreateNode`](https://github.com/mvisonneau/cilium/blob/e2676483a51b7b089cd6a476b83038b9a1ec6169/pkg/aws/eni/instances.go#L68), and we can rely on this to be set because the instance ID is [always set when the resource is created](https://github.com/mvisonneau/cilium/blob/e2676483a51b7b089cd6a476b83038b9a1ec6169/pkg/nodediscovery/nodediscovery.go#L469).

Something like:

```
func (m *InstancesManager) CreateNode(obj *v2.CiliumNode, n *ipam.Node) ipam.NodeOperations {
	return &Node{k8sObj: obj, manager: m, node: n, instanceID: n.InstanceID()}
}
```

or feel free to create a constructor for `Node`. 

If you think this is a good idea, then please implement in a separate commit. WDYT?

mvisonneau replied:
indeed, this is much cleaner and future proof, thanks for the guidance on this @christarazi 🙇 !

christarazi commented on pkg/ipam/node.go:
```diff
@@ -828,7 +831,7 @@ func (n *Node) update(origNode, node *v2.CiliumNode, attempts int, status bool)
 // For now, this function is only needed in ENI IPAM mode. In the future, this
 // may also be needed for Azure IPAM as well.
-func adjustPreAllocateIfNeeded(node *v2.CiliumNode) {
+func adjustPreAllocateIfNeeded(node *v2.CiliumNode) (err error) {
```
Nit: no need to name return values; we only have one

mvisonneau replied:
I reverted this change as well 👍

christarazi commented on pkg/ipam/node.go:
```diff
@@ -837,13 +840,22 @@ func adjustPreAllocateIfNeeded(node *v2.CiliumNode) {
+			`unable to find the limits for this instance type '%s',
+			consider setting --update-ec2-adapter-limit-via-api=true`,
+			node.Spec.ENI.InstanceType,
+		)
```
At this point in the code, this error can be triggered many times because the caller of this function will retry. Given that, it would be nicer if we used a `sync.Once` to log this only once, as to not pollute the logs. We can also provide the recommendation of setting the specific flag, rather than embedding the mitigation in the error msg. The error returned should simply be `errLimitsNotFound` or whatever the sentinel error variable is called.

We should also consider whether it makes more sense for the caller to log once or whether to do it in this function itself.

mvisonneau replied:
absolutely, I wasn't too sure about this part anyway. I believe it is out of scope for the PR so I will revert it for now.

mvisonneau replied:
done 👍

christarazi reviewed (changes requested):
Thanks for the PR! I think the core idea of the PR seems good to me. I have comments on how it's implemented however. 

Also, the commit msg seems to be better suited as a release note. The commit msg should briefly describe the code change, to conform to https://docs.cilium.io/en/v1.9/contributing/development/contributing_guide/#submitting-a-pull-request. Concretely, the commit msg can be:

```
ipam/aws: Provide mitigation if ENI limits not found
```

Elaborating that it fixed a bug would be good to move to the commit msg context instead.

Feel free to add / modify it as you see fit as well.

mvisonneau reviewed (commented):
thanks for the 👀 feedback @christarazi ! Can you please doublecheck a couple of my subsequent comments? 🙇

christarazi commented on pkg/nodediscovery/nodediscovery.go:
```diff
@@ -525,6 +525,8 @@ func determineFirstInterfaceIndex(instanceType string) *int {
 			idx = 0 // Include eth0
 		}
+	} else {
+		log.Warningf("Unable to find limits for instance type '%s', consider setting --update-ec2-adapter-limit-via-api=true", instanceType)
```
Rather than using dynamic string formatting, let's use structured logging instead:

```
log.WithField("instance-type", instanceType).Warning(...)
```

christarazi replied:
Actually, we might need string formatting to include the name of the option:

```
Warningf("Unable to find limits for instance type, consider setting --%s=true", option.UpdateEC2AdapterLimitViaAPI)
```

mvisonneau replied:
absolutely, I splitted the change in 2 commits and updated accordingly 👌

christarazi reviewed (changes requested):
Looks much better! Could you separate out the `instanceID` change into a separate commit? The log msg should be in it's own commit. This way we have clear logical commits implementing independent behavior.

mvisonneau committed:
nodediscovery: advise to set a flag when instance limits are not found

When instance limits are not found, we will now advise the enduser to
set --update-ec2-adapter-limit-via-api=true to allow the operator to
dynamically fetch the limits from the EC2 API

Signed-off-by: Maxime VISONNEAU <maxime.visonneau@gmail.com>

mvisonneau committed:
ipam/aws: provide mitigation if ENI limits not found

This change ensures that we do not endup with a deadlock of
the goroutines when the EC2 instance limits are not available.

The initial lock starts in pkg/ipam/node.go where we set
a "n.mutex.Lock()" in "func (n *Node) recalculate()".

If the EC2 instance type limits cannot be determined, in cilium/pkg/aws/eni/node.go,
there is a call to "n.loggerLocked()" being made which subsequently
calls "n.node.InstanceID()".

However, this function also expects to have a lock over the same mutex:
"n.mutex.RLock()" in pkg/ipam/node.go

Leading to a deadlock of the go routine, preventing the operator
from functioning in a nominal fashion for any instance of the cluster.

Fixes #14802

Signed-off-by: Maxime VISONNEAU <maxime.visonneau@gmail.com>

christarazi commented on pkg/nodediscovery/nodediscovery.go:
```diff
@@ -525,6 +526,13 @@ func determineFirstInterfaceIndex(instanceType string) *int {
+		}).Warningf(
+			"Unable to find limits for instance type, consider setting --%s=true",
+			operatorOption.UpdateEC2AdapterLimitViaAPI,
+		)
```
I didn't realize that the flag is only for the operator...

Maybe we should clarify that in the log msg. Something like:

> "Unable to find limits for instance type, consider setting --%s=true on the Operator"

mvisonneau replied:
sounds good, updated the log message and the commit message 👍

christarazi reviewed (commented):
Awesome! I have 2 more pieces of feedback:

- Minor fixup to the log msg
- Could you update the commit msg of the second commit to elaborate a bit more on the deadlock issue? Maybe describe that in a particular function (I forgot the name of it), a write-lock was obtained and while it was locked, calling `n.loggerLocked()` would call `n.node.InstanceID()` which then attempts to read-lock to extract out the `instanceID`.

christarazi commented on pkg/nodediscovery/nodediscovery.go:
```diff
@@ -525,6 +526,13 @@ func determineFirstInterfaceIndex(instanceType string) *int {
+		log.WithFields(logrus.Fields{
+			"instance-type": instanceType,
+		}).Warningf(
+			"Unable to find limits for instance type, consider setting --%s=true on the operator",
```
Capitalize "operator"

mvisonneau replied:
great, sorted! thanks for the very constructive review 🙇 !

christarazi reviewed (commented):
:tada: 

LGTM, thanks for the PR! Minor feedback:

- Minor nit below
- In the 2nd commit msg, I'd avoid referencing exact line numbers as they'll get out-of-date quickly. Referencing the function is good enough like you have already done. :)

christarazi reviewed (approved):
:rocket:
